tracing/traced_file.cc

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/future.hh>
+#include <seastar/core/print.hh>
+
+#include "tracing/traced_file.hh"
+
+using namespace seastar;
+
+class traced_file_impl : public file_impl {
+private:
+    // Invariant: _f_impl == *get_file_impl(_f).
+    // The _f is only stored to keep ownership.
+    file _f;
+    file_impl& _f_impl;
+
+    tracing::trace_state_ptr _trace_state;
+    sstring _trace_prefix;
+
+    template <typename... Args>
+    void trace(const char* fmt, Args&&... args) {
+        tracing::trace(_trace_state, "{} {}", _trace_prefix, format(fmt, std::forward<Args>(args)...));
+    }
```

Comment:
Why a nested format? Seems wasteful. tracing::trace() will also call format().

Reply:
I did it to prepend the message with `_trace_prefix`.

Another way would be to prepend it to `fmt`, which would require creating a temporary `sstring` variable for `_trace_prefix + " " + fmt`, which is more verbose.

Reply:
Prepending to fmt is dangerous, because files always have "{}" in them (okay, our database files don't, but this class shouldn't depend on it).

I think that printing _trace_prefix explicitly in all the call sites is better. Yes it's ugly, but it's only ugly at compile time.

Reply:
https://www.youtube.com/watch?v=WWaLxFIVX1s

Reply:
https://www.youtube.com/watch?v=07So_lJQyqw